src/Storages/MergeTree/MutateTask.cpp

```diff
@@ -292,13 +298,15 @@ getColumnsForNewDataPart(
     SerializationInfoByName new_serialization_infos;
     for (const auto & [name, old_info] : serialization_infos)
     {
-        if (removed_columns.contains(name))
-            continue;
-
         auto it = renamed_columns_from_to.find(name);
         auto new_name = it == renamed_columns_from_to.end() ? name : it->second;
 
-        if (!updated_header.has(new_name))
+        if (!storage_columns_set.contains(new_name))
+            continue;
+
+        /// In compact part we read all columns and all of them are in @updated_header.
+        /// But in wide part we must keep serialization infos for columns that are not touched by mutation.
+        if (!updated_header.has(new_name) && isWidePart(source_part))
```

Comment:
The issue is that sometimes column may be dropped implicitly by another mutation (see https://github.com/ClickHouse/ClickHouse/blob/c2a5318df4a91678850cf8f493e75728264cf139/src/Storages/MergeTree/MutateTask.cpp#L158-L164). Previously it wasn't accounted in serialization infos.